interpreter/golang/golang.go

```diff
@@ -0,0 +1,201 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package golang // import "go.opentelemetry.io/ebpf-profiler/interpreter/golang"
+
+/*
+#cgo CFLAGS: -g -Wall
+#include "../../rust-crates/symblib-capi/c/symblib.h"
+#include <stdlib.h>
+*/
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+	"sync/atomic"
+	"unsafe"
+
+	lru "github.com/elastic/go-freelru"
+	"go.opentelemetry.io/ebpf-profiler/host"
+	"go.opentelemetry.io/ebpf-profiler/interpreter"
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/metrics"
+	"go.opentelemetry.io/ebpf-profiler/process"
+	"go.opentelemetry.io/ebpf-profiler/remotememory"
+	"go.opentelemetry.io/ebpf-profiler/reporter"
+	"go.opentelemetry.io/ebpf-profiler/successfailurecounter"
+	"go.opentelemetry.io/ebpf-profiler/tpbase"
+)
+
+var (
+	// compiler check to make sure the needed interfaces are satisfied
+	_ interpreter.Data     = &golangData{}
+	_ interpreter.Instance = &golangInstance{}
+)
+
+// goSymData caches source information.
+type goSymData struct {
+	function   string
+	sourceFile string
+	line       libpf.SourceLineno
+}
+
+type golangData struct {
+	exec string
+}
+
+type golangInstance struct {
+	// Golang symbolization metrics
+	successCount atomic.Uint64
+	failCount    atomic.Uint64
+
+	// pcToFunc is a helper cache ... tdb
+	pcToFunc *lru.SyncedLRU[libpf.FrameID, goSymData]
+
+	goRuntime *C.SymblibPointResolver
+	pin       runtime.Pinner
+}
+
+func Loader(_ interpreter.EbpfHandler, info *interpreter.LoaderInfo) (
+	interpreter.Data, error) {
+	ef, err := info.GetELF()
+	if err != nil {
+		return nil, err
+	}
+	if !ef.IsGolang() {
+		return nil, nil
+	}
+
+	exec := info.FileName()
+	if len(exec) < 2 {
+		// There are cases where FileName() just returns /.
+		// In these cases we can not access the backing executable
+		// by FileName() and therefore can not continue here.
+		return nil, nil
+	}
+
+	return &golangData{
+		exec: exec,
+	}, nil
+}
+
+func (g *golangData) Attach(_ interpreter.EbpfHandler, pid libpf.PID,
+	_ libpf.Address, _ remotememory.RemoteMemory) (interpreter.Instance, error) {
+	pcToFunc, err := lru.NewSynced[libpf.FrameID, goSymData](1024, libpf.FrameID.Hash32)
+	if err != nil {
+		return nil, err
+	}
+	gi := &golangInstance{
+		pcToFunc: pcToFunc,
+	}
+
+	executablePath := C.CString(fmt.Sprintf("/proc/%d/root%s", pid, g.exec))
+	defer C.free(unsafe.Pointer(executablePath))
+
+	//nolint:gocritic
+	status := C.symblib_goruntime_new(executablePath, &gi.goRuntime)
+	if status != C.SYMBLIB_OK {
+		return nil, fmt.Errorf("failed to create point resolver for '%s': %d",
+			C.GoString(executablePath), status)
+	}
+	gi.pin.Pin(unsafe.Pointer(&gi.goRuntime))
+
+	return gi, nil
+}
+
+// SynchronizeMappings is a NOP for Golang.
+func (g *golangInstance) SynchronizeMappings(_ interpreter.EbpfHandler, _ reporter.SymbolReporter,
```

Comment:
These empty functions can go away if you embed `interpreter.InstanceStubs`.

Reply:
Resolving with https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/408/commits/256ee691a030e6e8ce10c2d55c266e86034c1e68